// @contentstack/delivery-sdk/dist/legacy/persistance/storages/local-storage.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/persistance/storages/local-storage.ts
var local_storage_exports = {};
__export(local_storage_exports, {
  localStorage: () => localStorage
});
module.exports = __toCommonJS(local_storage_exports);
var import_utils = require("../helper/utils.cjs");
var localStorage = {
  name: "localStorage",
  clear,
  each,
  getItem,
  removeItem,
  setItem
};
function _localStorage() {
  return import_utils.iGlobal.localStorage;
}
function clear() {
  _localStorage().clear();
}
function each(callback) {
  for (let i = _localStorage().length - 1; i >= 0; i--) {
    const key = _localStorage().key(i);
    if (key && getItem(key)) {
      const value = getItem(key);
      callback(value, key);
    }
  }
}
function setItem(key, value) {
  if (!key) {
    return;
  }
  _localStorage().setItem(key, value);
}
function getItem(key) {
  return _localStorage().getItem(key);
}
function removeItem(key) {
  _localStorage().removeItem(key);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  localStorage
});
