// @contentstack/delivery-sdk/dist/modern/persistance/storages/memory-storage.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/persistance/storages/memory-storage.ts
var memory_storage_exports = {};
__export(memory_storage_exports, {
  memoryStorage: () => memoryStorage
});
module.exports = __toCommonJS(memory_storage_exports);
var memoryStorage = {
  name: "memoryStorage",
  clear,
  each,
  getItem,
  removeItem,
  setItem
};
var memory = {};
function clear() {
  memory = {};
}
function each(callback) {
  for (const key in memory) {
    const value = getItem(key);
    callback(value, key);
  }
}
function setItem(key, value) {
  if (!key) {
    return;
  }
  memory[key] = value;
}
function getItem(key) {
  return memory[key];
}
function removeItem(key) {
  delete memory[key];
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  memoryStorage
});
