// @contentstack/delivery-sdk/dist/legacy/lib/entries.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/entries.ts
var entries_exports = {};
__export(entries_exports, {
  Entries: () => Entries
});
module.exports = __toCommonJS(entries_exports);
var import_core = require("@contentstack/core");
var import_query = require("./query.cjs");
var import_base_query = require("./base-query.cjs");
var import_utils = require("./utils.cjs");
var import_error_messages = require("./error-messages.cjs");
var Entries = class extends import_base_query.BaseQuery {
  constructor(client, contentTypeUid) {
    super();
    this._client = client;
    this._contentTypeUid = contentTypeUid;
    this._urlPath = `/content_types/${this._contentTypeUid}/entries`;
    this._variants = "";
  }
  /**
   * @method except
   * @memberof Entries
   * @description Excludes specific field/fields of an entry
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType("contentTypeUid").entry().except("fieldUID").find()
   *
   * @param {string} fieldUid - field uid to exclude
   * @returns {Entries} - returns Entries object for chaining method calls
   */
  except(fieldUid) {
    if (Array.isArray(fieldUid)) {
      let i = 0;
      for (const uid of fieldUid) {
        this._queryParams[`except[BASE][${i}]`] = uid;
        i++;
      }
    } else {
      this._queryParams["except[BASE][]"] = fieldUid;
    }
    return this;
  }
  /**
   * @method includeBranch
   * @memberof Entries
   * @description Includes the branch in result
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType(contentType_uid).entry().includeBranch().find();
   */
  includeBranch() {
    this._queryParams.include_branch = "true";
    return this;
  }
  /**
   * @method includeContentType
   * @memberof Entries
   * @description Include the details of the content type along with the entries details
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType(contentType_uid).entry().includeContentType().find();
   */
  includeContentType() {
    this._queryParams.include_content_type = "true";
    return this;
  }
  /**
   * @method includeEmbeddedItems
   * @memberof Entries
   * @description Include Embedded Objects (Entries and Assets) along with entry/entries details.
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType(contentType_uid).entry().includeEmbeddedItems().find();
   */
  includeEmbeddedItems() {
    this._queryParams["include_embedded_items[]"] = "BASE";
    return this;
  }
  /**
   * @method includeFallback
   * @memberof Entries
   * @description When an entry is not published in a specific language, content can be fetched from its fallback language
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType(contentType_uid).entry().includeFallback().find();
   */
  includeFallback() {
    this._queryParams.include_fallback = "true";
    return this;
  }
  /**
   * @method includeMetadata
   * @memberof Entries
   * @description Include the metadata for getting metadata content for the entry.
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType(contentType_uid).entry().includeMetadata().find();
   */
  includeMetadata() {
    this._queryParams.include_metadata = "true";
    return this;
  }
  /**
   * @method includeReference
   * @memberof Entries
   * @description To include the content of the referred entries in your response,
   * you need to use the include[] parameter and specify the UID of the reference field as value.
   * This function sets the include parameter to a reference field UID in the API request.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const entries = stack.contentType("contentTypeUid").entry().includeReference("brand")
   * const result = await entries.find()
   *
   * @param {string} referenceFieldUid - UID of the reference field to include.
   * @returns {Entries} - Returns the Entries instance for chaining.
   */
  includeReference(...referenceFieldUid) {
    if (referenceFieldUid.length) {
      referenceFieldUid.forEach((value) => {
        if (!Array.isArray(this._queryParams["include[]"])) {
          this._queryParams["include[]"] = [];
        }
        this._queryParams["include[]"].push(...Array.isArray(value) ? value : [value]);
      });
    } else {
      console.error(import_error_messages.ErrorMessages.INVALID_ARGUMENT_STRING_OR_ARRAY);
    }
    return this;
  }
  /**
   * @method includeReferenceContentTypeUID
   * @memberof Entries
   * @description This method also includes the content type UIDs of the referenced entries returned in the response.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const entries = stack.contentType("contentTypeUid").entry().includeReferenceContentTypeUID()
   * const result = await entries.find()
   *
   * @returns {Entries} - Returns the Entries instance for chaining.
   */
  includeReferenceContentTypeUID() {
    this._queryParams.include_reference_content_type_uid = "true";
    return this;
  }
  /**
   * @method includeSchema
   * @memberof Entries
   * @description Includes the schema of the content type along with the entries details.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const entries = stack.contentType("contentTypeUid").entry().includeSchema()
   * const result = await entries.find()
   *
   * @returns {Entries} - Returns the Entries instance for chaining.
   */
  includeSchema() {
    this._queryParams.include_schema = "true";
    return this;
  }
  /**
   * @method locale
   * @memberof Entries
   * @description The entries published in the locale will be fetched
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType("contentTypeUid").entry().locale('en-us').find();
   */
  locale(locale) {
    this._queryParams.locale = locale;
    return this;
  }
  /**
   * @method only
   * @memberof Entries
   * @description Selects specific field/fields of an entry
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType("contentTypeUid").entry().only("fieldUID").find()
   *
   * @param {string} fieldUid - field uid to select
   * @returns {Entries} - returns Entries object for chaining method calls
   */
  only(fieldUid) {
    if (Array.isArray(fieldUid)) {
      let i = 0;
      for (const uid of fieldUid) {
        this._queryParams[`only[BASE][${i}]`] = uid;
        i++;
      }
    } else {
      this._queryParams["only[BASE][]"] = fieldUid;
    }
    return this;
  }
  /**
   * @method query
   * @memberof Entries
   * @description Creates a query object for filtering entries
   * @param {object} queryObj - Optional query object to initialize the query
   * @returns {Query} Query instance for chaining query methods
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const query = stack.contentType("contentTypeUid").entry().query();
   */
  query(queryObj) {
    if (queryObj) return new import_query.Query(this._client, this._parameters, this._queryParams, this._variants, this._contentTypeUid, queryObj);
    return new import_query.Query(this._client, this._parameters, this._queryParams, this._variants, this._contentTypeUid);
  }
  /**
   * @method variants
   * @memberof Entries
   * @description The variant header will be added to axios client
   * @returns {Entries}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.contentType('abc').entry().variants('xyz').find();
   */
  variants(variants) {
    if (Array.isArray(variants) && variants.length > 0) {
      this._variants = variants.join(",");
    } else if (typeof variants == "string" && variants.length > 0) {
      this._variants = variants;
    }
    return this;
  }
  /**
   * Override find method to include content type UID directly for better caching
   */
  async find(encode = false) {
    let requestParams = this._queryParams;
    if (Object.keys(this._parameters).length > 0) {
      let queryParams = { ...this._parameters };
      if (encode) {
        queryParams = (0, import_utils.encodeQueryParams)(queryParams);
      }
      requestParams = { ...this._queryParams, query: queryParams };
    }
    const getRequestOptions = {
      params: requestParams,
      // Add contentTypeUid directly for improved caching
      contentTypeUid: this._contentTypeUid
    };
    if (this._variants) {
      getRequestOptions.headers = {
        ...getRequestOptions.headers,
        "x-cs-variant-uid": this._variants
      };
    }
    const response = await (0, import_core.getData)(this._client, this._urlPath, getRequestOptions);
    return response;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Entries
});
