// @contentstack/delivery-sdk/dist/modern/lib/asset.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/asset.ts
var asset_exports = {};
__export(asset_exports, {
  Asset: () => Asset
});
module.exports = __toCommonJS(asset_exports);
var import_core = require("@contentstack/core");
var Asset = class {
  constructor(client, assetUid) {
    this._queryParams = {};
    this._client = client;
    this._urlPath = `/assets/${assetUid}`;
  }
  /**
   * @method includeFallback
   * @memberof Asset
   * @description When an entry is not published in a specific language, content can be fetched from its fallback language
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').includeFallback().fetch();
   */
  includeFallback() {
    this._queryParams.include_fallback = "true";
    return this;
  }
  /**
   * @method includeMetadata
   * @memberof Asset
   * @description Include the metadata for getting metadata content for the asset.
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').includeMetadata().fetch();
   */
  includeMetadata() {
    this._queryParams.include_metadata = "true";
    return this;
  }
  /**
   * @method includeDimension
   * @memberof Asset
   * @description Includes the dimensions (height and width) of the image in result
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').includeDimension().fetch();
   */
  includeDimension() {
    this._queryParams.include_dimension = "true";
    return this;
  }
  /**
   * @method includeBranch
   * @memberof Asset
   * @description Includes the branch in result
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').includeBranch().fetch();
   */
  includeBranch() {
    this._queryParams.include_branch = "true";
    return this;
  }
  /**
   * @method relativeUrls
   * @memberof Asset
   * @description Includes the relative URLs of the asset in result
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').relativeUrls().fetch();
   */
  relativeUrls() {
    this._queryParams.relative_urls = "true";
    return this;
  }
  /**
   * @method version
   * @memberof Asset
   * @description Retrieve a specific version of an asset in result
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').version(1).fetch();
   */
  version(version) {
    this._queryParams.version = String(version);
    return this;
  }
  /**
   * @method locale
   * @memberof Asset
   * @description The asset published in the locale will be fetched
   * @returns {Asset}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').locale('en-us').fetch();
   */
  locale(locale) {
    this._queryParams.locale = locale;
    return this;
  }
  /**
   * @method fetch
   * @memberof Asset
   * @description Fetches the asset data on the basis of the asset uid
   * @returns {Promise<T>} Promise that resolves to the asset data
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset('asset_uid').fetch();
   */
  async fetch() {
    const response = await (0, import_core.getData)(this._client, this._urlPath, { params: this._queryParams });
    if (response.asset) return response.asset;
    return response;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Asset
});
