// @contentstack/delivery-sdk/dist/modern/lib/asset-query.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/asset-query.ts
var asset_query_exports = {};
__export(asset_query_exports, {
  AssetQuery: () => AssetQuery
});
module.exports = __toCommonJS(asset_query_exports);
var import_base_query = require("./base-query.cjs");
var import_query = require("./query.cjs");
var AssetQuery = class extends import_base_query.BaseQuery {
  constructor(client) {
    super();
    this._client = client;
    this._urlPath = "/assets";
  }
  /**
   * @method version
   * @memberof AssetQuery
   * @description Retrieve a specific version of an asset in result
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().version(1).find();
   */
  version(version) {
    this._queryParams.version = String(version);
    return this;
  }
  /**
   * @method includeDimension
   * @memberof AssetQuery
   * @description Includes the dimensions (height and width) of the image in result
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().includeDimension().find();
   */
  includeDimension() {
    this._queryParams.include_dimension = "true";
    return this;
  }
  /**
   * @method includeBranch
   * @memberof AssetQuery
   * @description Includes the branch in result
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().includeBranch().find();
   */
  includeBranch() {
    this._queryParams.include_branch = "true";
    return this;
  }
  /**
   * @method includeMetadata
   * @memberof AssetQuery
   * @description Include the metadata for getting metadata content for the asset.
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().includeMetadata().find();
   */
  includeMetadata() {
    this._queryParams.include_metadata = "true";
    return this;
  }
  /**
   * @method relativeUrls
   * @memberof AssetQuery
   * @description Includes the relative URLs of the assets in result
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().relativeUrls().find();
   */
  relativeUrls() {
    this._queryParams.relative_url = "true";
    return this;
  }
  /**
   * @method includeFallback
   * @memberof AssetQuery
   * @description When an entry is not published in a specific language, content can be fetched from its fallback language
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().includeFallback().find();
   */
  includeFallback() {
    this._queryParams.include_fallback = "true";
    return this;
  }
  /**
   * @method locale
   * @memberof AssetQuery
   * @description The assets published in the locale will be fetched
   * @returns {AssetQuery}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().locale('en-us').find();
   */
  locale(locale) {
    this._queryParams.locale = locale;
    return this;
  }
  /**
   * @method query
   * @memberof AssetQuery
   * @description Fetches the asset data on the basis of the query
   * @returns {Query}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * const result = await stack.asset().query().where('fieldUid', QueryOperation.EQUALS, 'value').find();
   */
  query() {
    return new import_query.Query(this._client, this._parameters, this._queryParams);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AssetQuery
});
