// @contentstack/delivery-sdk/dist/modern/lib/stack.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/stack.ts
var stack_exports = {};
__export(stack_exports, {
  Stack: () => Stack
});
module.exports = __toCommonJS(stack_exports);
var import_core = require("@contentstack/core");
var import_asset = require("./asset.cjs");
var import_asset_query = require("./asset-query.cjs");
var import_content_type = require("./content-type.cjs");
var import_contenttype_query = require("./contenttype-query.cjs");
var import_synchronization = require("./synchronization.cjs");
var import_taxonomy_query = require("./taxonomy-query.cjs");
var import_global_field_query = require("./global-field-query.cjs");
var import_global_field = require("./global-field.cjs");
var Stack = class {
  constructor(client, config) {
    this._client = client;
    this.config = config;
    this._client.stackConfig = this.config;
  }
  asset(uid) {
    if (uid) return new import_asset.Asset(this._client, uid);
    return new import_asset_query.AssetQuery(this._client);
  }
  contentType(uid) {
    if (uid) return new import_content_type.ContentType(this._client, uid);
    return new import_contenttype_query.ContentTypeQuery(this._client);
  }
  /**
     * @method Taxonomy
     * @memberof Stack
     * @description Sets the url to /taxonomies/entries. Pass a query to fetch entries with taxonomies
     *
     * @returns {TaxonomyQuery}
     * @example
     * import contentstack from '@contentstack/delivery-sdk'
     *
     * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
  
     * const taxonomy = stack.taxonomy() // For taxonomy query object
     */
  taxonomy() {
    return new import_taxonomy_query.TaxonomyQuery(this._client);
  }
  globalField(uid) {
    if (uid) return new import_global_field.GlobalField(this._client, uid);
    return new import_global_field_query.GlobalFieldQuery(this._client);
  }
  /**
   * @method setLocale
   * @memberof Stack
   * @description Sets the locale of the API server
   * @param {String} locale - valid locale e.g. fr-fr
   * @return {Stack}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * stack.setLocale('en-155');
   */
  setLocale(locale) {
    this.config.locale = locale;
    this._client.defaults.params.locale = locale;
  }
  /**
   * @method sync
   * @memberof Stack
   * @description Syncs your Contentstack data with your app and ensures that the data is always up-to-date by providing delta updates
   * @param {object} params - params is an object that supports 'locale', 'start_date', 'content_type_uid', and 'type' queries.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync()        // For initializing sync
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'locale': 'en-us'})     //For initializing sync with entries of a specific locale
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'start_date': '2018-10-22'})    //For initializing sync with entries published after a specific date
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'content_type_uid': 'session'})   //For initializing sync with entries of a specific content type
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'type': 'entry_published'})
   * //Use the type parameter to get a specific type of content. Supports'asset_published',
   * // 'entry_published', 'asset_unpublished', 'entry_unpublished', 'asset_deleted', 'entry_deleted', 'content_type_deleted'.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({'pagination_token': '<page_tkn>'})    // For fetching the next batch of entries using pagination token
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({'sync_token': '<sync_tkn>'})    // For performing subsequent sync after initial sync
   * @instance
   */
  async sync(params = {}, recursive = false) {
    return await (0, import_synchronization.synchronization)(this._client, params, recursive);
  }
  livePreviewQuery(query) {
    if (this.config.live_preview) {
      let livePreviewParams = { ...this.config.live_preview };
      if (query.live_preview) {
        livePreviewParams = {
          ...livePreviewParams,
          live_preview: query.live_preview,
          contentTypeUid: query.contentTypeUid || query.content_type_uid,
          entryUid: query.entryUid || query.entry_uid,
          preview_timestamp: query.preview_timestamp || "",
          include_applied_variants: query.include_applied_variants || false
        };
      } else {
        livePreviewParams = {
          ...livePreviewParams,
          live_preview: "",
          contentTypeUid: "",
          entryUid: "",
          preview_timestamp: "",
          include_applied_variants: false
        };
      }
      this._client.stackConfig.live_preview = livePreviewParams;
    }
    if (query.hasOwnProperty("release_id")) {
      this._client.defaults.headers["release_id"] = query.release_id;
    } else {
      delete this._client.defaults.headers["release_id"];
    }
    if (query.hasOwnProperty("preview_timestamp")) {
      this._client.defaults.headers["preview_timestamp"] = query.preview_timestamp;
    } else {
      delete this._client.defaults.headers["preview_timestamp"];
    }
  }
  getClient() {
    return this._client;
  }
  async getLastActivities() {
    try {
      const result = await (0, import_core.getData)(this._client, "/content_types", {
        params: {
          only_last_activity: true,
          environment: this.config.environment
        }
      });
      return result;
    } catch (error) {
      throw new Error("Error fetching last activities");
    }
  }
  /**
   * @method setPort
   * @memberof Stack
   * @description Sets the port of the host
   * @param {Number} port - Port Number
   * @return {Stack}
   * @instance
   * */
  setPort(port) {
    if (typeof port === "number") this.config.port = port;
    return this;
  }
  /**
   * @method setDebug
   * @memberof Stack
   * @description Sets the debug option
   * @param {boolean} debug - Debug value
   * @return {Stack}
   * @instance
   * */
  setDebug(debug) {
    if (typeof debug === "boolean") this.config.debug = debug;
    return this;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Stack
});
